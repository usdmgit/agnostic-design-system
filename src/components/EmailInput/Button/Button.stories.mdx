# Button Component

## Type: Simple

## Button Props

OBS: An asterisk (\*) denotes that the property will have additional info below the table.

|        Prop        |                     Type                      |                                                                                                                                                                  Description                                                                                                                                                                  |                                                           Default                                                           | Mandatory |
| :----------------: | :-------------------------------------------: | :-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------: | :-------------------------------------------------------------------------------------------------------------------------: | :-------: |
|     appendIcon     |                React.ReactNode                | It is possible to send your own [Icon](https://design-system.codelitt.dev/?path=/story/components-button--with-custom-append-icon). The Icon can also be the [only element inside the button](https://design-system.codelitt.dev/?path=/story/components-button--without-label-with-append-icon), if there isn't a value passed to the label. | By default it receives an [Arrow Icon](https://design-system.codelitt.dev/?path=/story/components-button--with-append-icon) |   false   |
|     category\*     |                   Category                    |                                                                                                                     A button can have one of any six categories (neutral, primary, secondary, positive, negative, ghost).                                                                                                                     |                                                           neutral                                                           |   false   |
|      content       |                React.ReactNode                |                                                                                                                                 Expects a JSX block that is implicitly return and rendered inside the button.                                                                                                                                 |                                                              -                                                              |   false   |
|      disabled      |                    boolean                    |                                                                                                                                           Specifies whether or not a button is disabled/actionable.                                                                                                                                           |                                                            false                                                            |   false   |
|   iconPosition\*   |                   Position                    |                                                                                                                                                   Position where the Icon will be displayed                                                                                                                                                   |                                                            right                                                            |   false   |
|         id         |                    string                     |                                                                                                                                                                  Unique id.                                                                                                                                                                   |                                                              -                                                              |   false   |
|       fixed        |                    boolean                    |                                                                                                                Passes the following styling, if set to **true**, to the button: max-width: 100%; min-width: 100%; width: 100%;                                                                                                                |                                                            false                                                            |   false   |
|      onClick       |                  () => void                   |                                                                                                                                               Calls a function whenever the button is clicked.                                                                                                                                                |                                                              -                                                              |   true    |
|       size\*       |                     Size                      |                                                                                                                    Size of the component. The change is reflected on the height and width. Can be either large or medium.                                                                                                                     |                                                            large                                                            |   false   |
|        text        |                    string                     |                                                                                                                                                     Text that displays inside the button.                                                                                                                                                     |                                                              -                                                              |   false   |
|        type        | 'button' \| 'submit' \| 'reset' \| undefined; |                                                                                                                                            Reflects the type attribute of the html button element.                                                                                                                                            |                                                          'button'                                                           |   false   |
| variablesClassName |                    string                     |                                                                                                                                                        Accepts custom CSS class names.                                                                                                                                                        |                                                              -                                                              |   false   |
|   withAppendIcon   |                    boolean                    |                                                                                                                                       Specifies whether or not a button should display its appendIcon.                                                                                                                                        |                                                            false                                                            |   false   |

---

**Categories availables:**

1.  [neutral](https://design-system.codelitt.dev/?path=/story/components-button--neutral)
2.  [primary](https://design-system.codelitt.dev/?path=/story/components-button--primary)
3.  [secondary](https://design-system.codelitt.dev/?path=/story/components-button--secondary)
4.  [positive](https://design-system.codelitt.dev/?path=/story/components-button--positive)
5.  [negative](https://design-system.codelitt.dev/?path=/story/components-button--negative)
6.  [ghost](https://design-system.codelitt.dev/?path=/story/components-button--ghost)

**Sizes**

1. large
2. medium

**Positions**

1. left
2. right

## Getting Started

```javascript
import React, { Component } from 'react';
import { Button } from '@codelittinc/agnostic-design-system';

class SimpleButton extends Component {
  const handleClick = () => {
    console.log('Button clicked!');
  }

  render() {
    return (
      <Button onClick={handleClick} />
    );
  }
}

export default SimpleButton;

```

## Variables available

|  Property   |    Attribute     |             State              |
| :---------: | :--------------: | :----------------------------: |
|             |   align-items    |                                |
|             |  align-content   |                                |
|             | background-color | hover, focus, active, disabled |
|             |      border      | hover, focus, active, disabled |
|             |      bottom      |                                |
|             |  border-radius   |                                |
|             |    box-shadow    | hover, focus, active, disabled |
|             |      color       | hover, focus, active, disabled |
|             |      cursor      | hover, focus, active, disabled |
|             |     display      |                                |
|             |  flex-direction  |                                |
|             |    flex-wrap     |                                |
|             |   font-family    |                                |
|             |    font-size     |                                |
|             |   font-weight    |                                |
|             |      height      |                                |
|             | justify-content  |                                |
|             |   line-height    |                                |
|             |      margin      |                                |
|             |     opacity      |                                |
|             |     outline      |     hover, focus, active       |
|             |  outline-offset  |     hover, focus, active       |
|             |     padding      |                                |
|             |    position      |                                |
|             |      width       |                                |
| append-icon |      color       | hover, focus, active, disabled |
| append-icon |      margin      |                                |
|   content   |     overflow     |                                |
|   content   |  text-overflow   |                                |
|   content   |   white-space    |                                |
