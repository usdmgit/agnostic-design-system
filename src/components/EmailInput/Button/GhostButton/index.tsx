import React from 'react';
import DefaultButton from '@/components/Button/DefaultButton';
import styles from '@/components/Button/GhostButton/GhostButton.css';
import classNames from 'classnames';

interface Props {
  onClick?: () => void;
}

const GhostButton = React.forwardRef<HTMLButtonElement, Props>((props, ref) => (
  <DefaultButton
    {...props}
    className={classNames(styles.button, styles['button--ghost'])}
    ref={ref}
  />
));

export default GhostButton;
