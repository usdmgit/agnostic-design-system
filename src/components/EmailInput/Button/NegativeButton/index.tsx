import React from 'react';
import DefaultButton from '@/components/Button/DefaultButton';
import styles from '@/components/Button/NegativeButton/NegativeButton.css';
import classNames from 'classnames';

interface Props {
  onClick?: () => void;
}

const NegativeButton = React.forwardRef<HTMLButtonElement, Props>((props, ref) => (
  <DefaultButton
    {...props}
    className={classNames(styles.button, styles['button--negative'])}
    ref={ref}
  />
));

export default NegativeButton;
