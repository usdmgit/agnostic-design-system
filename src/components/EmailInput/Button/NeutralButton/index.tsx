import React from 'react';
import DefaultButton from '@/components/Button/DefaultButton';
import styles from '@/components/Button/NeutralButton/NeutralButton.css';
import classNames from 'classnames';

interface Props {
  onClick?: () => void;
}

const NeutralButton = React.forwardRef<HTMLButtonElement, Props>((props, ref) => (
  <DefaultButton
    {...props}
    className={classNames(styles.button, styles['button--neutral'])}
    ref={ref}
  />
));

export default NeutralButton;
