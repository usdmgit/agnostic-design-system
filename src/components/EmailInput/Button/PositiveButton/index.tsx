import React from 'react';
import DefaultButton from '@/components/Button/DefaultButton';
import styles from '@/components/Button/PositiveButton/PositiveButton.css';
import classNames from 'classnames';

interface Props {
  onClick?: () => void;
}

const PositiveButton = React.forwardRef<HTMLButtonElement, Props>((props, ref) => (
  <DefaultButton
    {...props}
    className={classNames(styles.button, styles['button--positive'])}
    ref={ref}
  />
));

export default PositiveButton;
