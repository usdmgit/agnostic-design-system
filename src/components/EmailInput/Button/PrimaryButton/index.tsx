import React from 'react';
import DefaultButton from '@/components/Button/DefaultButton';
import styles from '@/components/Button/PrimaryButton/PrimaryButton.css';
import classNames from 'classnames';

interface Props {
  onClick?: () => void;
}

const PrimaryButton = React.forwardRef<HTMLButtonElement, Props>((props, ref) => (
  <DefaultButton
    {...props}
    className={classNames(styles.button, styles['button--primary'])}
    ref={ref}
  />
));

export default PrimaryButton;
