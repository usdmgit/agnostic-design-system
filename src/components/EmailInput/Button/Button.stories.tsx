import React from 'react';
import { Meta } from '@storybook/react';
import Button from '@/components/Button';
import TestSearchIcon from '@/assets/images/icons/web/search-simple.svg';
import mdx from './Button.stories.mdx';

export default {
  title: 'Components/Button',
  component: Button,
  parameters: {
    docs: {
      page: mdx
    }
  }
} as Meta;

const Template = args => <Button {...args} />;

export const Neutral = Template.bind({});
Neutral.args = {
  text: 'Neutral Button',
  category: 'neutral'
};

export const Primary = Template.bind({});
Primary.args = {
  text: 'Primary Button',
  category: 'primary'
};

export const Secondary = Template.bind({});
Secondary.args = {
  text: 'Secondary Button',
  category: 'secondary'
};

export const Positive = Template.bind({});
Positive.args = {
  text: 'Positive Button',
  category: 'positive'
};

export const Negative = Template.bind({});
Negative.args = {
  text: 'Negative Button',
  category: 'negative'
};

export const Ghost = Template.bind({});
Ghost.args = {
  text: 'Ghost Button',
  category: 'ghost'
};

export const WithAppendIcon = Template.bind({});
WithAppendIcon.args = {
  text: 'With Append Icon',
  category: 'primary',
  withAppendIcon: true
};

export const WithLeftAppendIcon = Template.bind({});
WithLeftAppendIcon.args = {
  text: 'With Left Append Icon',
  category: 'primary',
  iconPosition: 'left',
  withAppendIcon: true
};

export const WithRightAppendIcon = Template.bind({});
WithRightAppendIcon.args = {
  text: 'With Right Append Icon',
  category: 'primary',
  iconPosition: 'right',
  withAppendIcon: true
};

export const WithCustomAppendIcon = Template.bind({});
WithCustomAppendIcon.args = {
  text: 'With Custom Append Icon',
  category: 'primary',
  withAppendIcon: true,
  appendIcon: <TestSearchIcon height='10' width='10' />
};

export const WithoutLabelWithAppendIcon = Template.bind({});
WithoutLabelWithAppendIcon.args = {
  category: 'primary',
  withAppendIcon: true
};

export const WithCustomContent = Template.bind({});
WithCustomContent.args = {
  category: 'secondary',
  content: (
    <div>
      <TestSearchIcon />
      <div>Text</div>
    </div>
  )
};
