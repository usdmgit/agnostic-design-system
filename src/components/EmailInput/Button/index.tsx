import React from 'react';

import GhostButton from '@/components/Button/GhostButton';
import NegativeButton from '@/components/Button/NegativeButton';
import NeutralButton from '@/components/Button/NeutralButton';
import PositiveButton from '@/components/Button/PositiveButton';
import PrimaryButton from '@/components/Button/PrimaryButton';
import SecondaryButton from '@/components/Button/SecondaryButton';

export type Category = 'ghost' | 'negative' | 'neutral' | 'positive' | 'primary' | 'secondary';
export type Position = 'left' | 'right';
export type Size = 'large' | 'medium';

export const SUBMIT_TYPE = 'submit';

interface Props {
  appendIcon?: React.ReactNode;
  category?: Category;
  content?: React.ReactNode;
  disabled?: boolean;
  fixed?: boolean;
  iconPosition?: Position;
  id?: string;
  text?: string | React.ReactNode;
  onClick?: () => void;
  size?: Size;
  type?: 'button' | 'submit' | 'reset' | undefined;
  variablesClassName?: string;
  withAppendIcon?: boolean;
}

const buttons = {
  ghost: GhostButton,
  negative: NegativeButton,
  neutral: NeutralButton,
  positive: PositiveButton,
  primary: PrimaryButton,
  secondary: SecondaryButton
};

const Button = React.forwardRef<HTMLButtonElement, Props>((props, ref) => {
  const ButtonType = buttons[props.category || 'neutral'];

  return <ButtonType {...props} ref={ref} />;
});

export default Button;
