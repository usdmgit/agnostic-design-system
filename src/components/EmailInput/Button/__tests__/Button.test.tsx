import React from 'react';
import Button from '@/components/Button';
import { mount } from 'enzyme';

describe('ExampleComponent', () => {
  it('is renders correctly', () => {
    const wrapper = mount(<Button text='this is a test' onClick={jest.fn} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('is renders the variablesClassName', () => {
    const className = 'my-cool-class';
    const wrapper = mount(
      <Button variablesClassName={className} text='this is a test' onClick={jest.fn} />
    );
    expect(wrapper.props().variablesClassName).toEqual(className);
  });

  it('is renders a fixed button', () => {
    const wrapper = mount(<Button text='this is a test' onClick={jest.fn} fixed />);
    expect(wrapper.exists('.button--fixed')).toBeTruthy();
  });

  it('is renders a primary button', () => {
    const wrapper = mount(<Button text='this is a test' onClick={jest.fn} category='primary' />);
    expect(wrapper.exists('.button--primary')).toBeTruthy();
  });

  it('is renders a secondary button', () => {
    const wrapper = mount(
      <Button text='this is a test' onClick={jest.fn} category='secondary' fixed />
    );
    expect(wrapper.exists('.button--secondary')).toBeTruthy();
  });

  it('is renders a positive button', () => {
    const wrapper = mount(<Button text='this is a test' onClick={jest.fn} category='positive' />);
    expect(wrapper.exists('.button--positive')).toBeTruthy();
  });

  it('is renders a neutral button', () => {
    const wrapper = mount(<Button text='this is a test' onClick={jest.fn} />);
    expect(wrapper.exists('.button--neutral')).toBeTruthy();
  });

  it('is renders a negative button', () => {
    const wrapper = mount(<Button text='this is a test' onClick={jest.fn} category='negative' />);
    expect(wrapper.exists('.button--negative')).toBeTruthy();
  });

  it('is renders a ghost button', () => {
    const wrapper = mount(<Button text='this is a test' onClick={jest.fn} category='ghost' />);
    expect(wrapper.exists('.button--ghost')).toBeTruthy();
  });
});
