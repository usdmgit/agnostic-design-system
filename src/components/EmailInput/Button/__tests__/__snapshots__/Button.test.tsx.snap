// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExampleComponent is renders correctly 1`] = `
<ForwardRef
  onClick={[Function]}
  text="this is a test"
>
  <ForwardRef
    onClick={[Function]}
    text="this is a test"
  >
    <ForwardRef
      className="button button--neutral"
      disabled={false}
      fixed={false}
      iconPosition="right"
      onClick={[Function]}
      size="large"
      text="this is a test"
    >
      <button
        className="button button--neutral button button--large"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        <span
          className="button-text"
        >
          this is a test
        </span>
      </button>
    </ForwardRef>
  </ForwardRef>
</ForwardRef>
`;
